data/build_redial.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2020 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Preprocessing Scripts for the Redial Dataset."""
+import json
+from tqdm import tqdm
+import numpy as np
+import re
+import os
+
+def prepare_redial(data_dir):
+    """Processes raw redial data in data_dir and saves the results in data folder."""
+    print("--Loading Redial Dataset--")
+    train = read_jsonl(os.path.join(data_dir, "rd-train-data.jsonl"))
+    test = read_jsonl(os.path.join(data_dir,"rd-test-data.jsonl"))
+
+
+    print("--Replacing Movie IDs--")
+    for dialogue in tqdm(train):
+        replace_ids(dialogue)
+    for dialogue in tqdm(test):
+        replace_ids(dialogue)
+
+    print("-Formatting For Training--")
+    train_formatted = separate_responses(train)
+    test_formatted = separate_responses(test)
+
+    write_jsonl("./redial/rd-train-formatted.jsonl", train_formatted)
+    write_jsonl("./redial/rd-test-formatted.jsonl", test_formatted)
+
+def length_summary(data_dir):
+    """prints a five number summary of the lengths of redial input/outputs."""
+    print("--Loading Dataset For Summary--")
+    train_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+    test_data = read_jsonl(os.path.join(data_dir, "rd-train-formatted.jsonl"))
+
+    def len_function(key):
+        return lambda x: len(x[key].split())
+
+    lengths = {"train_inputs" : list(map(len_function("conversation"), train_data)),
+                "train_targets" : list(map(len_function("response"), train_data)),
+                "test_inputs" : list(map(len_function("conversation"), test_data)),
+                "test_targets" : list(map(len_function("response"), test_data))}
+
+    for name, length_array in lengths.items():
+        print(name.upper() + ": ")
+        quartile_summary(length_array)
+
+#Helper Functions
+
+def read_jsonl(filename):
+    """Reads a jsonl file and returns as array of dicts."""
```

Comment:
array of dicts or list of dicts? Not sure if there is a difference in python. :)

Reply:
List of dicts you're right. I'm used to treating them like arrays. Changed the comment.